src/main/kotlin/snyk/common/lsp/progress/ProgressManager.kt

```diff
@@ -0,0 +1,235 @@
+        }
+
+    fun isDisposed() = disposed
+
+    init {
+        Disposer.register(SnykPluginDisposable.getInstance(), this)
+    }
+
+    fun createProgress(params: WorkDoneProgressCreateParams): CompletableFuture<Void> {
+        if (!disposed) {
+            val token = getToken(params.token)
+            getProgress(token)
+        }
+        return CompletableFuture.completedFuture(null)
+    }
+
+    private fun createProgressIndicator(progress: Progress) {
+        val token: String = progress.token
+        if (isDone(progress)) {
+            progresses.remove(token)
+            return
+        }
+        val title = "Snyk: " + progress.title
+        val cancellable: Boolean = progress.cancellable
+        ProgressManager.getInstance()
+            .run(newProgressBackgroundTask(title, cancellable, progress, token))
+    }
+
+    private fun newProgressBackgroundTask(
+        title: String,
+        cancellable: Boolean,
+        progress: Progress,
+        token: String
+    ): Task.Backgroundable {
+        val project = ProjectUtil.getActiveProject()
+        return object : Task.Backgroundable(project, title, cancellable) {
+            override fun run(indicator: ProgressIndicator) {
+                try {
+                    while (!isDone(progress)) {
+                        if (indicator.isCanceled) {
+                            cancelProgress(token)
+                            throw ProcessCanceledException()
+                        }
+
+                        var progressNotification: WorkDoneProgressNotification?
+                        try {
+                            progressNotification = progress.nextProgressNotification
+                        } catch (e: InterruptedException) {
+                            progresses.remove(token)
+                            Thread.currentThread().interrupt()
+                            throw ProcessCanceledException(e)
+                        }
+                        if (progressNotification != null) {
+                            val kind = progressNotification.kind ?: return
+                            when (kind) {
+                                WorkDoneProgressKind.begin ->  // 'begin' has been notified
+                                    begin(progressNotification as WorkDoneProgressBegin, indicator)
+
+                                WorkDoneProgressKind.report ->  // 'report' has been notified
+                                    report(progressNotification as WorkDoneProgressReport, indicator)
+
+                                WorkDoneProgressKind.end -> Unit
+                            }
+                        }
+                    }
+                } finally {
+                    indicator.cancel()
+                    progresses.remove(token)
+                }
+            }
+        }
+    }
+
+    private fun cancelProgress(token: String) {
+        try {
+            progresses[token]?.cancel()
```

Comment:
am wondering if we need a to a mutex here

Reply:
progresses is a concurrentMap - so we don't